include/unifex/filter_stream.hpp

```diff
@@ -0,0 +1,250 @@
+
+  template(typename CPO)                       //
+      (requires is_receiver_query_cpo_v<CPO>)  //
+      friend auto tag_invoke(CPO cpo, const type& r) noexcept(
+          is_nothrow_callable_v<CPO, const Receiver&>)
+          -> callable_result_t<CPO, const Receiver&> {
+    return std::move(cpo)(std::as_const(r.op_.receiver_));
+  }
+
+#if UNIFEX_ENABLE_CONTINUATION_VISITATIONS
+  template <typename Func>
+  friend void
+  tag_invoke(tag_t<visit_continuations>, const type& r, Func&& func) {
+    std::invoke(func, r.op_.receiver_);
+  }
+#endif
+
+  template <typename... Values>
+  void set_value(Values... values) && noexcept {
+    auto& op = op_;
+    op.next_.destruct();
+    UNIFEX_TRY {
+      const bool doFilter =
+          !std::invoke(op.filter_, std::forward<Values>(values)...);
+
+      if (doFilter) {
+        op.next_.construct_with([&] {
+          return unifex::connect(next(op.stream_), next_receiver_t{op});
+        });
+        unifex::start(op.next_.get());
+      } else {
+        unifex::set_value(std::move(op.receiver_), std::move(values)...);
+      }
+    }
+    UNIFEX_CATCH(...) {
+      unifex::set_error(std::move(op.receiver_), std::current_exception());
+    }
+  }
+
+  void set_done() && noexcept {
+    auto& op = op_;
+    op.next_.destruct();
+    unifex::set_done(std::move(op.receiver_));
+  }
+
+  void set_error(std::exception_ptr ex) && noexcept {
+    auto& op = op_;
+    op.next_.destruct();
+    unifex::set_error(std::move(op.receiver_), std::move(ex));
+  }
+
+  template <typename Error>
+  void set_error(Error&& e) && noexcept {
+    std::move(*this).set_error(make_exception_ptr(std::forward<Error>(e)));
+  }
+};
+
+template <typename StreamSender, typename FilterFunc, typename Receiver>
+struct _op<StreamSender, FilterFunc, Receiver>::type {
+  StreamSender& stream_;
+  FilterFunc filter_;
+  Receiver receiver_;
+
+  using next_receiver_t = next_receiver<StreamSender, FilterFunc, Receiver>;
+  using next_op =
+      manual_lifetime<next_operation_t<StreamSender, next_receiver_t>>;
+
+  next_op next_;
+
+  template <typename StreamSender2, typename FilterFunc2, typename Receiver2>
+  explicit type(
+      StreamSender2&& stream, FilterFunc2&& filter, Receiver2&& receiver)
+    : stream_(std::forward<StreamSender2>(stream))
+    , filter_(std::forward<FilterFunc2>(filter))
+    , receiver_(std::forward<Receiver2>(receiver)) {}
+
+  // Question to @ispeters: This didn't make a lot of sense to me, but now that
+  // we don't have a union, do we want to remove this?
+  ~type() {}  // Due to the union member, this is load-bearing. DO NOT DELETE.
+
+  void start() noexcept {
+    UNIFEX_TRY {
+      next_.construct_with([&] {
+        return unifex::connect(next(stream_), next_receiver_t{*this});
+      });
+      unifex::start(next_.get());
+    }
+    UNIFEX_CATCH(...) {
+      unifex::set_error(std::move(receiver_), std::current_exception());
+    }
+  }
+};
```

Comment:
I think you should connect the underlying stream's next-sender into `next_` in the operation state's constructor; that way, this type will have the invariant that `next_` always contains a fully-constructed value, which means this type's destructor can unconditionally destruct `next_`, which will, I think, simplify some lifetime concerns in the next-receiver.
```suggestion
  template <typename StreamSender2, typename FilterFunc2, typename Receiver2>
  explicit type(
      StreamSender2&& stream, FilterFunc2&& filter, Receiver2&& receiver)
      noexcept(
          std::is_nothrow_constructible_v<StreamSender, StreamSender2>
          && std::is_nothrow_constructible_v<FilterFunc, FilterFunc2>
          && std::is_nothrow_constructible_v<Receiver, Receiver2>
          && is_nothrow_connectable_v<next_sender_t<StreamSender>, Receiver&>)
    : stream_(std::forward<StreamSender2>(stream))
    , filter_(std::forward<FilterFunc2>(filter))
    , receiver_(std::forward<Receiver2>(receiver)) {
    next_.construct_with([&] {
      return unifex::connect(next(stream_), next_receiver_t{*this});
    });
  }

  ~type() {
    next_.destruct();
  }

  void start() noexcept {
    unifex::start(next_.get());
  }
};
```

Reply:
After commenting on the next-receiver about relying on this supposed invariant, I realize it's not maintained when `connect` can throw so we'll need a `bool` to keep track of whether `next_` has a value.
```suggestion
  bool nextEngaged_{false};

  template <typename StreamSender2, typename FilterFunc2, typename Receiver2>
  explicit type(
      StreamSender2&& stream, FilterFunc2&& filter, Receiver2&& receiver)
      noexcept(
          std::is_nothrow_constructible_v<StreamSender, StreamSender2>
          && std::is_nothrow_constructible_v<FilterFunc, FilterFunc2>
          && std::is_nothrow_constructible_v<Receiver, Receiver2>
          && is_nothrow_connectable_v<next_sender_t<StreamSender>, Receiver&>)
    : stream_(std::forward<StreamSender2>(stream))
    , filter_(std::forward<FilterFunc2>(filter))
    , receiver_(std::forward<Receiver2>(receiver)) {
    next_.construct_with([&] {
      return unifex::connect(next(stream_), next_receiver_t{*this});
    });
    nextEngaged_ = true;
  }

  ~type() {
    if (nextEngaged_) {
      next_.destruct();
    }
  }

  void start() noexcept {
    unifex::start(next_.get());
  }
};
```

Reply:
One thing I'm not seeing is when we connect to the next receiver & that potentially throws (assuming also the constructor is `noexcept(false)`), why don't we `set_error(...)` on the receiver?

Also with your version, it is not the case that `start` is `noexcept`.

Reply:
> One thing I'm not seeing is when we connect to the next receiver & that potentially throws (assuming also the constructor is `noexcept(false)`), why don't we `set_error(...)` on the receiver?

The receiver contract says that, once you've invoked `start()` on an operation state, exactly one of the receiver's completion methods will be invoked exactly once (with an exception for when `set_value` throws an exception). If the operation state's constructor throws an exception (i.e. if `connect` throws) then you haven't invoked `start()`, yet, so it's wrong to invoke any completion methods on the receiver.

> Also with your version, it is not the case that `start` is `noexcept`.

Why? `unifex::start(auto& op)` is required to be `noexcept` so `unifex::start(next_.get())` ought to be `noexcept`.

Reply:
Thanks for explaining :) I wasn't familiar with those, makes sense now.